### ISSUE TITLE
Subtasks for unaligned and event-data refactor

### ISSUE THREAD

**Issue by SimonHeybrock** [scipp/scipp#998]
This is a collection of independent tasks that could go into https://github.com/scipp/scipp/tree/natural-api. Coordination required!

Overall, the new API is introduced in the phases:

- [x] Phase 1: Remove old "sparse data" support. The consequence is that we will temporarily not be able to store multiple event data entries in a dataset. See #1001).

- [x] Phase 2: @SimonHeybrock (branch https://github.com/scipp/scipp/tree/unaligned and #1023) Add `unaligned` property to `DataArray`, referencing a nested unaligned array. (next after 1.)
     - The wrapping data array is refered to as "realigned" data array, since it realigns the unaligned data to new axis (lazily, i.e., no (re)binning happend)
     - include basic histogramming for demonstration and verification purposes.
     - apart from the implementation, there should be no difference between event data and other unaligned data. The only difference is that in the former case the unaligned data array is not aware of the "abstract" "event" dimension. On the realigned interface little difference might be noticable.
     - Support for realigned data in `neutron::convert`. Need to convert bins of realigned wrapper as wells as events in unaligned content.
     - Support for changing binning. Should this be done using `rebin`? Or just by setting new coord, with special checks to "resize" realigned content? `rebin` would make inplace changes weird, but setting coord resulting in a resize is also weird.

- [x] Phase 3: Basic realigned event data. See #1030.
     - [x] Slicing realigned just returns a view. Should probably not support modification of unaligned content via such a view, or (as a more complex alternative) be smart enough to apply operation only the selected events.
     - [x] Histogramming sliced realigned gives only the "filtered" events.
     - [x] Binary operations between normal dense data and realigned data. This is essentially the "new" interface for what we have already as operations between histograms and event data. Now it will additionally use the bin edge coords of the realigned array to check the coordinates before the operation.
     - [x] Support for realigned data in neutron::convert. Need to convert bins of realigned wrapper as wells as events in unaligned content.
     - [x] Support for changing binning. Should this be done using rebin? Or just by setting new coord, with special checks to "resize" realigned content? rebin would make inplace changes weird, but setting coord resulting in a resize is also weird.

- [ ] Phase 4: Add new features such as histogramming, slicing (event filtering), and more. This will be many sub-tasks than are partially independent and can probably be implemented in a relatively free timeline.
     - [x] Copying sliced realigned should actually filter out events that are out of bounds (keeping the events). (#1040)
     - [x] Event filtering (#1040)
     - [x] `realign` should actually also filter already. (#1040)
     - [ ] Simultaneous support for multiple unaligned dimensions and unaligned events.
          - [x] Multiple realigned event dims. (#1040)
          - [ ] Combinations of normal and event realignment
     - [x] Histogramming needs to be able to handle data in events lists, normal data (e.g., mapping from `position` to `x,y,z`), and combinations of the two.
          - [ ] 2D histogram for event data
          - [ ] Mixed normal + event data histogramming
     - [x] Binary operations and others should follow the **"as-if"** rule: **Any operation with realigned data should operate *as if* it had been done with *histogrammed* data**.
       - [ ] For binary operations this mostly says that histogramming before or after the operation should give the same result. For example, `histogram(realigned/dense) == histogram(realigned)/dense` (apart from the subtlety of uncertainties, see discussion a while back), whereas `histogram(dense/realigned) != dense/histogram(realigned)`. Another valid example is `histogram(realigned1 + realigned2) == histogram(realigned1) + histogram(realigned2)`, i.e., summing with the realigned interface would, e.g., concatenate event lists.
       - [ ] For unary operations this implies that the operation has to be **distributive over addition** (histogramming is addition). For example, `sin(histogram(realigned)) != histogram(sin(realigned))`, so it should not be supported.
       - [ ] Summing over realigned dimension would simply remove that coordinate (unless it is the last one). For example, this can be used to go from `\theta, \phi` to just `\theta`.
     - [ ] `groupby` could support realigned data, to provide advanced filtering based on aligned auxiliary coords.

- [x] Omitted above: Mask and attribute handling. This is similar but slightly different from coord handling. I am not sure yet whether this should be part of phase 2 or 3, or be done later.

Independent subtasks after #1023 and/or #1030:
- [x] Add support for "realigned" data arrays in `to_string`, `__repr__`, and `to_html`. Should the html via have a way to expand and show the `unaligned` data array? -> Not done, see #1150.
- [x] Adapt `plot` to allow for on-the-fly histogramming. In particular the slice viewer should support histogramming only the displayed slice, instead of requiring histogramming the full volume.
- [ ] Support event data for the case where events are *not* being realigned. Calling `histogram` in this case would, e.g., "bin" positions into x,y,z bins but preserve events. This would probably simply require a call to `groupby.flatten` instead of `groupby.sum` in `histogram_md_recurse` if elements are event lists.
- [ ] Make use of `realign` in instrument view?

Related cleanup tasks see #1047.

**Comment by SimonHeybrock** [comment]
#798 may be covered by this if we implement multi-dimensional histogramming.

**Comment by SimonHeybrock** [comment]
Done or superseded by event-data rewrite.
